keps/sig-cluster-lifecycle/wgs/2141-instance-specific-componentconfig/README.md

```diff
@@ -0,0 +1,664 @@
+6. Decode flags over `SharedInternal` and `InstanceInternal` to enforce flag precedence.
+   Flags for fields that exist in both types must be registered against `InstanceInternal`,
+   so that we maintain flag precedence when we merge the two.
+7. Copy fields that exist in both types from `InstanceInternal` to `SharedInternal` so that
+   the values from `InstanceInternal` are present everywhere.
+
+In the diagram below it's possible to have a better visualization of the case-by-case merge flow:
+
+![component-case-merge](images/component-case-merge.png)
+
+### Why not just keep using flags for instance-specific parameters?
+
+Flags have various problems, outlined in the
+[Versioned Component Configuration Files](https://docs.google.com/document/d/1FdaEJUEh091qf5B98HM6_8MS764iXrxxigNIdwHYW9c)
+doc.
+
+Leaving some parameters as flags while the rest are exposed via ComponentConfig
+results in an inconsistent API surface, and an inconsistent API versioning
+policy. Why fix only part of the problem when we can fix the whole thing?
+
+### Why not just auto-detect?
+
+One alternative that has been proposed in the past is to find a way to
+auto-detect values of instance-specific parameters and simply eliminate them
+from the configuration workflow. This is a great idea in theory, because it
+solves the problem by removing work for users, but it may be difficult in
+practice. For example, on a machine with multiple network cards, which IP should
+the kubelet use?
+
+This solution is orthogonal and does not prevent us from finding ways to
+auto-detect instance specific config in the future. It is worth noting that
+the _possibility_ of someday being able to auto-detect these has led to some
+hesitation to make a firm decision today. We have delayed for too long already,
+ and a firm decision is needed to move forward.
+
+### Risks and Mitigations
+
+For users deploying Pod templates, this solution can make the template more
+verbose. In the future, we can explore more elegant ways of generating files
+with instance-specific parameters correctly substituted. This proposal argues
+that the extra verbosity is worth the benefit of a consistent, versioned API for
+configuring core components.
+
+## Design Details
+
+The canonical flag for providing a ComponentConfig file to a component is
+`--config`. This KEP proposes that the canonical flag for providing an
+instance-specific config file be `--instance-config`.
+
+As with sharable ComponentConfig parameters, command line flags for
+instance-specific parameters should continue to function and take precedence
+over the final merged values from the config file so that backward compatibility
+is maintained.
+
+### Configuration merge example
+
+The merge between both YAML configuration files is provided by the
+[Strategic Merge Patch](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/strategic-merge-patch.md)
```

Comment:
this library is dead and going to be replaced by server-side-apply

Reply:
the SMP looked like the best approach since it was possible to configure field struct tags and was still being used internally in some parts. 

Thanks for informing the deprecation, I will rewrite the KEP using SSA instead.